akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,545 @@
+          if (activeConsumers == 0) {
+            if (readyToShutdown) completeStage()
+            else if (head < finalOffset) {
+              // If our final consumer goes away, we roll forward the buffer so a subsequent consumer does not
+              // see the already consumed elements. This feature is quite handy.
+              while (head < finalOffset) {
+                queue(head & Mask) = null
+                head += 1
+              }
+              head = finalOffset
+              if (!hasBeenPulled(in)) pull(in)
+            }
+          } else checkUnblock(previousOffset)
+        case Advance(id, previousOffset) ⇒
+          val newOffset = previousOffset + DemandThreshold
+          debug(s"advance consumer $id from offset $previousOffset to $newOffset ")
+          val consumer = findAndRemoveConsumer(id, previousOffset)
+          addConsumer(consumer, newOffset)
+          checkUnblock(previousOffset)
+        case NeedWakeup(id, previousOffset, currentOffset) ⇒
+          val consumer = findAndRemoveConsumer(id, previousOffset)
+          addConsumer(consumer, currentOffset)
+          debug(s"wakup request $id, previous $previousOffset, current: $currentOffset, tail is $tail")
+
+          if (currentOffset == tail) {
+            debug(s"$id wants wakeup, but we can't help")
+          } else {
+            debug(s"$id is behind us, wake it up")
+            consumer.callback.invoke(Wakeup)
+          }
+          try { checkUnblock(previousOffset) } catch { case ex ⇒ ex.printStackTrace() }
+      }
+    }
+
+    // Producer API
+    private def isBlocked: Boolean = tail - head == bufferSize
+
+    override def onUpstreamFailure(ex: Throwable): Unit = {
+      val failMessage = HubCompleted(Some(ex))
+
+      // Notify pending consumers and set tombstone
+      state.getAndSet(Closed(Some(ex))).asInstanceOf[Open].registrations foreach { consumer ⇒
+        consumer.callback.invoke(failMessage)
+      }
+
+      // Notify registered consumers
+      consumerWheel.iterator.flatMap(_.iterator) foreach { consumer ⇒
+        consumer.callback.invoke(failMessage)
+      }
+      failStage(ex)
+    }
+
+    private def findAndRemoveConsumer(id: Long, offset: Int): Consumer = {
+      // TODO: Try to eliminate modulo division somehow...
+      val wheelSlot = offset % (bufferSize + 1)
+      var consumersInSlot = consumerWheel(wheelSlot)
+      //debug(s"consumers before removal $consumersInSlot")
+      var remainingConsumersInSlot: List[Consumer] = Nil
+      var removedConsumer: Consumer = null
+
+      while (consumersInSlot.nonEmpty) {
+        val consumer = consumersInSlot.head
+        if (consumer.id != id) remainingConsumersInSlot = consumer :: remainingConsumersInSlot
+        else removedConsumer = consumer
+        consumersInSlot = consumersInSlot.tail
+      }
+      consumerWheel(wheelSlot) = remainingConsumersInSlot
+      removedConsumer
+    }
+
+    private def unblockIfPossible(offset: Int): Boolean = {
```

Comment:
rename param to indicate that this is the offset of a consumer we just removed from that slot